### PR TITLE
Begin Fastlane setup with Match

### PR DESCRIPTION
**Pull request by designatednerd**
I created a `Fastfile` skeleton and then started filling stuff in around `match`. So far, only one lane is fully working and there's a massive number of `# TODO`s, but this is a place where we can start. 

NOTE: Once this PR is merged (or to test it before merging), you need to do ~two~ three things to continue being able to run on device while developing without having to monkey around with the `.xcodeproj` vfile: 
- Message me for the `local_config.sh` file
- run `bundle install` to make sure you have all the appropriate ruby nonsense installed for fastlane, bundler, and match
- run `bundle exec fastlane match_fetch_all_the_things` to execute the one working lane, which will fetch the signing info from a private repo. 

**Update, 1/2/18:** Match is now working properly and is being used to run tests on Circle on this PR.

### CODE REVIEW

**Comment by designatednerd** [comment]
The shell script only affects running with Fastlane, not within Xcode. Once you've run `match` on your local machine you should be able to just build and submit with the installed cert and profile if you want/need to do it without Fastlane.

**Comment by Elland** [comment]
@designatednerd so it has no effect on new users trying to run the project from their local machines? 👍🏽good to know.

**Comment by designatednerd** [comment]
@Elland the only effect it would have is if you're running on device you need to make sure you have the correct dev certificate.

File: Toshi/Controllers/Chat/ChatViewController.swift [hunk 1]
```diff
@@ -180,7 +180,7 @@ final class ChatViewController: UIViewController, UINavigationControllerDelegate
         super.viewWillDisappear(animated)
 
         isVisible = false
-        heightOfKeyboard = 0
+        keyboardScreenPositionOffset = 0
```
Comment (Elland) [inline Toshi/Controllers/Chat/ChatViewController.swift]: What's wrong with master, that these changes appear in all open PRs?

Comment (yuliaveres) [reply Toshi/Controllers/Chat/ChatViewController.swift]: @Elland due to broken input that fix was reverted before urgent release.
if we need this change, we might want to add fix for hidden input as well and have it in PR again

Comment (Elland) [reply Toshi/Controllers/Chat/ChatViewController.swift]: I did add a fix for the broken input. 👍 https://github.com/toshiapp/toshi-ios-client/pull/641/commits/7055d0b567c5d7fafa3f585e2e4d949dd12645f7

Comment (yuliaveres) [reply Toshi/Controllers/Chat/ChatViewController.swift]: ok. we will find time to qa and check it 👍  thank you !

**Comment by designatednerd** [comment]
@Elland @yuliaveres when you get back from vacay this is ready for a fresh look

File: Toshi.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -3566,7 +3566,8 @@
 			buildSettings = {
 				ASSETCATALOG_COMPILER_APPICON_NAME = "AppIcon-Distribution";
 				CODE_SIGN_ENTITLEMENTS = Toshi.entitlements;
-				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
+				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Distribution: Toshi Holdings PTE LTD (BH32JXCQWU)";
```
Comment (Elland) [inline Toshi.xcodeproj/project.pbxproj]: This looks a bit odd. From developer to distribution identity?

File: Toshi/Controllers/Browse/BrowseViewController.swift [hunk 3]
```diff
@@ -126,6 +126,32 @@ class BrowseViewController: SearchableCollectionController {
         searchField?.backgroundColor = Theme.inputFieldBackgroundColor
         
         addSubviewsAndConstraints()
+
+        _ = NotificationCenter.default.addObserver(forName: .UIKeyboardWillShow, object: nil, queue: .main) { [weak self] notification in
```
Comment (Elland) [inline Toshi/Controllers/Browse/BrowseViewController.swift]: Would be great to get these changes fixed back into master before merging other PRs, right?

File: fastlane/Fastfile [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    # Happier setup with Match and keychain.
+    setup_circle_ci
+
+    # Make sure Fastlane is using UTF-8
+    ENV["LC_ALL"]="en_US.UTF-8"
+    ENV["LANG"]="en_US.UTF-8"
+
+    if !is_ci
+      # Open-source note: This file is kept private to developers who can release.
+      sh ". ./local_config.sh"
```
Comment (Elland) [inline fastlane/Fastfile]: Is this right? `. ./local_config.sh` Not `../`?

File: Toshi.xcodeproj/project.pbxproj (hunk 2 above)
Comment (designatednerd) [reply Toshi.xcodeproj/project.pbxproj]: This closure is the app store release configuration - I'm not entirely sure why it wasn't the distribution identity before.

File: fastlane/Fastfile (hunk 4 above)
Comment (designatednerd) [reply fastlane/Fastfile]: Yeah, it's a weird thing about executing a shell script that `export`s a bunch of stuff into `env`. I still don't 100% understand it but I've been using it for like 4 years so 🤷‍♀️

File: Toshi/Controllers/Browse/BrowseViewController.swift (hunk 3 above)
Comment (designatednerd) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: I'll see if I can remove these changes - IIRC the changes are on the `feature/group_chat` branch, right, @yuliaveres?

Comment (yuliaveres) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: those were reverted and they are not on group chat branch

Comment (designatednerd) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: @yuliaveres should I pull them out to a separate PR or just delete them?

Comment (yuliaveres) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: you should delete them, afaik, @Elland has a different pr for that

Comment (designatednerd) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: @Elland which PR?

Comment (Elland) [reply Toshi/Controllers/Browse/BrowseViewController.swift]: https://github.com/toshiapp/toshi-ios-client/pull/641

**Comment by yuliaveres** [comment]
i have a lot of alerts `codesign wants to aceess key "N2X3A86PLS" in your keychain`. what is the password to that certificate ?

**Comment by yuliaveres** [comment]
but in general i managed to execute `bundle exec fastlane match_fetch_all_the_things` successfully

**Comment by designatednerd** [comment]
so the password being prompted for there is your keychain password - the cert is already installed on your computer, Xcode is just attempting to access it. Note that the prompt is asking for the password to your `login` keychain.

**Comment by Elland** [comment]
How is it looking? Working fine?

**Comment by designatednerd** [comment]
So far so good but I want to hold off on merging this until we get the current version shipped

**Comment by designatednerd** [comment]
Gonna be some signififcant changes to how we do this, will redo and reopen.

**Pull request closed without merging**